python/sparkdl/transformers/named_image.py

```diff
@@ -140,17 +141,31 @@ def __call__(self, value):
 _scaleHintConverter = _LazyScaleHintConverter()
 
 
-class DeepImageFeaturizer(JavaTransformer, HasInputCol, HasOutputCol):
+class _DeepImageFeaturizerReader(JavaMLReader):
+    def __init__(self, clazz):
+        super(_DeepImageFeaturizerReader, self).__init__(clazz)
+
+    @classmethod
+    def _java_loader_class(cls, clazz):
+        return "com.databricks.sparkdl.DeepImageFeaturizer"
+
+
+class DeepImageFeaturizer(JavaTransformer, JavaMLReadable, JavaMLWritable):
     """
     Applies the model specified by its popular name, with its prediction layer(s) chopped off,
     to the image column in DataFrame. The output is a MLlib Vector so that DeepImageFeaturizer
     can be used in a MLlib Pipeline.
     The input image column should be ImageSchema.
     """
-
+    # NOTE: We are not inheriting from HasInput/HasOutput to mirror the scala side. It also helps
+    # us to avoid issue with serialization/deserialization - default values based on uid do not
+    # always get set to correct value on the python side after deserialization. Default values do
+    # not get reset to the jvm side value unless they param value is not set.
+    # See pyspark.ml.wrapper.JavaParams._transfer_params_from_java
```

Comment:
btw, is this really broken in Spark (maybe they have some workarounds)? wouldn't the output column param be very broken for all the transformers? or do some of them go through different code paths that avoid this error? ???

Reply:
i think i do remember seeing some of the persistent code that reads in each param specifically from java to python so those probably avoid the issue. mostly trying to understand if there are any actual Transformers that have persistent broken currently in MLlib.

Reply:
I think most of them do. It only matters if someone queries default params directly though 
 so it's probably not a top priority. Here's the repro with bucketizer:
```
from pyspark.ml.feature import Bucketizer
a = Bucketizer()
a.save("bucketizer0")
b = load("bucketizer0")
 a._defaultParamMap[a.outputCol]
u'Bucketizer_4b489ecc5cae3e86faf6__output'
b._defaultParamMap[b.outputCol]
u'Bucketizer_42388aff94842ab55f2f__output'
```

Reply:
To make things worse:
```
 a.isSet(a.outputCol)
False
b.isSet(b.outputCol)
True
```

Reply:
I've created a Jira for the issue https://issues.apache.org/jira/browse/SPARK-23244

Reply:
Awesome, thanks!